Add tracing feature

aon opened the pull request:
Creates a trait with a `with_tracing` method to instrument a handler with a custom `tracing::Span`

WaffleLapkin commented on crates/teloxide/src/dispatching/dispatcher.rs:
```diff
@@ -495,6 +495,32 @@ impl<R, Err, Key> Dispatcher<R, Err, Key> {
+#[cfg(feature = "tracing")]
+pub trait UpdateHandlerTracingExt<E> {
+    /// Returns an `UpdateHandler` with tracing enabled.
+    fn with_tracing(self, f: fn(Arc<Update>) -> tracing::Span) -> Self;
```
```suggestion
    fn with_tracing(self, f: impl Fn(DependencyMap) -> tracing::Span) -> Self;
```

Two things here:
1. I think providing `DependencyMap` is better, as it is more flexible
2. You generally want to use `Fn` trait bound and not a function pointer

WaffleLapkin replied:
Or maybe using the `Injectable` trait from `dptree` is even better (esp given the typeck PR to `dptree` that could make use of the types in the signature)

aon replied:
I'm not sure where the `Injectable` trait would go, I'm not too familiar with `dptree`. Would you care to explain your reasoning here?

aon replied:
I have a working version using `DependencyMap`, although I don't quite like the fact that I have to clone it in order to pass it down. Do you think it could affect performance? I'm not sure how other way it could be done.

```rust
// Removed comments for easier read.
pub trait UpdateHandlerTracingExt<E> {
    fn with_tracing_span<F>(self, f: F) -> Self
    where
        F: Fn(DependencyMap) -> tracing::Span + Send + Sync + 'static;
}

impl<E: 'static> UpdateHandlerTracingExt<E> for UpdateHandler<E> {
    fn with_tracing_span<F>(self, f: F) -> Self
    where
        F: Fn(DependencyMap) -> tracing::Span + Send + Sync + 'static,
    {
        let description = DpHandlerDescription::entry().merge_chain(self.description());

        dptree::from_fn_with_description(description, move |deps: DependencyMap, cont| {
            self.clone().execute(deps.clone(), cont).instrument(f(deps))
        })
    }
}
```

WaffleLapkin commented on crates/teloxide/src/dispatching/dispatcher.rs:
```diff
@@ -495,6 +495,32 @@ impl<R, Err, Key> Dispatcher<R, Err, Key> {
+
+#[cfg(feature = "tracing")]
+impl<E: 'static> UpdateHandlerTracingExt<E> for UpdateHandler<E> {
+    fn with_tracing(self, f: fn(Arc<Update>) -> tracing::Span) -> Self {
```
I would also prefer the method to have `span` in the name, something like `with_tracing_span` or maybe even just `spanned` although maybe it's ambiguous.

aon replied:
Great, changing it to `with_tracing_span`

WaffleLapkin commented on crates/teloxide/src/dispatching/dispatcher.rs:
```diff
@@ -495,6 +495,32 @@ impl<R, Err, Key> Dispatcher<R, Err, Key> {
+
+        // This is a hacky replacement for `handler.description().clone()`.
+        // Ideally cloning `DpHandlerDescription` would be supported by `teloxide`.
+        let description = DpHandlerDescription::entry().merge_chain(self.description());
```
Please leave a `FIXME` comment here, so we can fix this later.

aon replied:
Added it.

WaffleLapkin commented on crates/teloxide/src/dispatching/dispatcher.rs:
```diff
@@ -495,6 +495,32 @@ impl<R, Err, Key> Dispatcher<R, Err, Key> {
+///
+/// This trait is only available if the `tracing` feature is enabled.
+#[cfg(feature = "tracing")]
+pub trait UpdateHandlerTracingExt<E> {
```
This should be in its own module, not `dispatching::dispatcher`, but `dispatching::tracing`

aon replied:
Moved to new module.

WaffleLapkin commented:
> @WaffleLapkin I'm not sure if I should also modify CHANGELOG.md as well, let me know.

Yes, please update the changelog.

aon commented:
@WaffleLapkin I made the changes we discussed:

- Modified `CHANGELOG.md`
- Added FIXME comment
- Changed `with_tracing` to `with_tracing_span`
- Changed `with_tracing_span` function to receive argument with trait bound and not function pointer
- Changed `with_tracing_span` to pass `DependencyMap` to the user-provided function that returns the span (didn't do it using `Injectable` as I couldn't understand how that could be integrated here, let me know if I should change it)
- Moved to its own module

The only thing I'm kinda concerned is that passing `DependencyMap` forces to `clone()` it in order to both execute and pass it to the span function. I'm not sure if this could hit performance or if there's any other way it could be done.

WaffleLapkin commented on crates/teloxide/src/dispatching/tracing.rs:
```diff
@@ -0,0 +1,29 @@
+    /// Returns an `UpdateHandler` with tracing enabled.
+    fn with_tracing_span<F>(self, f: F) -> Self
+    where
+        F: Fn(DependencyMap) -> tracing::Span + Send + Sync + 'static;
```
To use `Injectable` the bound should be something like this:
```rust
F: Injectable<DependencyMap, tracing::Span, FnArgs>
```

Where `FnArgs` is a newly added generic parameter. (btw you should probably import `tracing::Span`)

Then you can use it like this:
```rust
f.inject(&deps)()
```

aon replied:
I tried implementing it with `Injectable` but the issue I'm having is that the `.inject()` returns a `Future` that I have to await in order to get `tracing::Span`, but `dptree::from_fn_with_description` doesn't expect an async closure.

For reference, this is the code I'm testing with:

```rust
pub trait UpdateHandlerTracingExt<E> {
    fn with_tracing_span<F, FnArgs>(self, f: F) -> Self
    where
        F: Injectable<DependencyMap, Span, FnArgs>;
}

impl<E: 'static> UpdateHandlerTracingExt<E> for UpdateHandler<E> {
    fn with_tracing_span<F, FnArgs>(self, f: F) -> Self
    where
        F: Injectable<DependencyMap, Span, FnArgs>,
    {
        let description = DpHandlerDescription::entry().merge_chain(self.description());

        dptree::from_fn_with_description(description, move |deps: DependencyMap, cont| async {
            let span = f.inject(&deps)().await;
            self.clone().execute(deps, cont).instrument(span)
        })
    }
}

```

WaffleLapkin replied:
You can simply add an `.await` after the instrument and it should work:
```rust
self.clone().execute(deps, cont).instrument(span).await
```

WaffleLapkin commented:
@aon I've lost track of the status here, is it waiting on you, or is it waiting on review/someone else?

aon commented:
It was waiting for me. Sorry about that, I couldn't make it work at the time and forgot to update the PR. It's now working, I changed it to use `Injectable`.

aon commented:
For usage reference, it can be used like this:

```rust
let dispatcher = Dispatcher::builder(
    bot,
    entry_handler.with_tracing_span(|deps: DependencyMap| async move {
        let update: Arc<Update> = deps.get();
        tracing::info_span!(
            "Handle update",
            username = update.from().and_then(|user| user.username.clone())
        )
    }),
)
.dependencies(dptree::deps![InMemStorage::<State>::new(), google])
.enable_ctrlc_handler()
.build();
```

WaffleLapkin commented on crates/teloxide/src/dispatching/tracing.rs:
```diff
@@ -0,0 +1,34 @@
+
+        dptree::from_fn_with_description(description, move |deps: DependencyMap, cont| {
+            let self_c = self.clone();
+            let f_c = f.clone();
```
I don't think you need this `.clone()` (or the `Clone` bound), `inject` already take a `&`.

aon replied:
Yeah this is basically why I'm doing it:

![image](https://github.com/teloxide/teloxide/assets/21188659/ac7f85af-4dd8-428d-b0a8-5ca521779cf8)

I couldn't find another way to solve this.

WaffleLapkin replied:
Try doing
```diff
-let span = f.inject(&deps)().await;
+let f = f.inject(&deps);
+let span = f().await;
+drop(f);
```

Cause I think this should be equivalent to what other dptree things do: https://github.com/teloxide/dptree/blob/da3de9d583380cfe2192f10d1d7681c3e79339a1/src/handler/map.rs#L83-L85

aon replied:
I see, there you're using an `Arc` to allow `Fn` to capture the variable. Makes sense, this works now, added the changes to the PR.

WaffleLapkin commented on crates/teloxide/src/dispatching/tracing.rs:
```diff
@@ -0,0 +1,34 @@
+
+/// Extension methods for adding `tracing` to [`UpdateHandler`].
+///
+/// This trait is only available if the `tracing` feature is enabled.
```
This is not needed, rustdoc generates such messages automatically.

WaffleLapkin commented on crates/teloxide/src/dispatching/tracing.rs:
```diff
@@ -0,0 +1,34 @@
+///
+/// This trait is only available if the `tracing` feature is enabled.
+pub trait UpdateHandlerTracingExt<E> {
+    /// Returns an `UpdateHandler` with tracing enabled.
```
You can't disable tracing... This doesn't enable it, this just wraps everything in a span.

WaffleLapkin commented on crates/teloxide/src/dispatching/tracing.rs:
```diff
@@ -0,0 +1,34 @@
+/// This trait is only available if the `tracing` feature is enabled.
+pub trait UpdateHandlerTracingExt<E> {
+    /// Returns an `UpdateHandler` with tracing enabled.
+    fn with_tracing_span<F, FnArgs>(self, f: F) -> Self
```
Can you rename this to `instrument_with_async` + add `instrument_with` which is the same, but uses `Asyncify<F>: ` bound? (see the `dptree` for how async/sync versions are implemented).

WaffleLapkin commented:
@teloxidebot author
(when ready for review, use `@teloxidebot ready`)

aon commented:
@teloxidebot ready

WaffleLapkin reviewed (commented):
This basically looks good to me. Although I'm wandering if we should move all of this to the `dptree` crate directly, after all there isn't anything Telegram or `teloxide` specific here.

@Hirrolot what do you think?

WaffleLapkin commented:
(fyi CI is failing because we set it up improperly and it allowed us to merge a PR that failed CI; #950 should fix the failing)

WaffleLapkin commented:
Note: this requires a `git rebase master` to get the CI fix

WaffleLapkin commented:
@Hirrolot I think this is waiting on your opinion

madetara commented:
Any chance for this to be picked up? I'm interested in this feature

I can also help if that's needed but it seems almost finished

Hirrolot commented:
@madetara, we were busy with releasing a new version, so this PR got somewhat delayed. I'll review it ASAP so that we could ship it in the next version.

shdwchn10 commented:
@hirrolot If you don't mind merge this into `teloxide`, I can do it

hirrolot commented:
@shdwchn10, yes, please do if it looks good.
